After branching on node type use uncheckedDowncast<>

annevk opened the pull request:
#### d8978483a98776f1e51225731fcb8d74c62d605a
<pre>
After branching on node type use uncheckedDowncast&lt;&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=268751">https://bugs.webkit.org/show_bug.cgi?id=268751</a>

Reviewed by Chris Dumez.

We can use the riskier uncheckedDowncast&lt;&gt; when it is clear the type is
correct, as it is in these switch statements.

* Source/WebCore/dom/Range.cpp:
(WebCore::processContentsBetweenOffsets):
(WebCore::Range::checkNodeOffsetPair):
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::appendNonElementNode):
* Source/WebCore/editing/MarkupAccumulator.h:
(WebCore::MarkupAccumulator::endAppendingNode):

Adopt dynamicDowncast&lt;&gt; while here.

Canonical link: <a href="https://commits.webkit.org/274207@main">https://commits.webkit.org/274207@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67a759089a6e735b9de5703fca36871a096ea067

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17225 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40597 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40839 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34053 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40527 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19985 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14564 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38854 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/14546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12659 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34222 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42117 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/34846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38477 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13218 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36675 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14773 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8614 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14237 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -681,16 +683,16 @@ ExceptionOr<Ref<DocumentFragment>> Range::cloneContents()
-    bool startIsText = startContainerNodeType == Node::TEXT_NODE;
-    if (startIsText && !startContainer().parentNode())
+
+    RefPtr startTextNode = dynamicDowncast<Text>(startContainer());
```
I *think*  this is not equivalent and we have to be careful. `startContainerNodeType == Node::TEXT_NODE` would return true for a Text node but not a CDataSection node. However, `dynamicDowncast<Text>(startContainer())` will be happy to cast a CDataSection node to a Text since it is one of its base classes.

annevk replied:
Thanks, I've reverted this and will tackle this separately as discussed. (It does seem like the DOM Standard wants us to be making this change, but it also seems there's no test coverage for it currently.)

annevk commented on Source/WebCore/editing/MarkupAccumulator.h:
```diff
@@ -127,8 +127,8 @@ class MarkupAccumulator {
-        appendEndTag(m_markup, downcast<Element>(node));
+    if (RefPtr element = dynamicDowncast<Element>(node))
+        appendEndTag(m_markup, *element);
     else if (auto element = replacementElement(node))
```
@cdumez `replacementElement()` is another method that returns RefPtr where it is not clear to me that it should per your earlier feedback. If you agree happy to tackle that in a follow-up.

cdumez replied:
`replacementElement()` is holding a RefPtr to the element internally on the stack. Therefore there is no cost to returning a RefPtr (no extra ref'ing). Yet, if the caller needs a RefPtr, returning a (free) RefPtr is beneficial. I'd keep `replacementElement()` returning a RefPtr.

However, I would replace `auto` with `RefPtr` here for clarity. It'd be nice if you could do this before landing.

annevk replied:
Will do and will do the same for the other caller.

annevk committed:
After branching on node type use uncheckedDowncast<>
https://bugs.webkit.org/show_bug.cgi?id=268751

Reviewed by Chris Dumez.

We can use the riskier uncheckedDowncast<> when it is clear the type is
correct, as it is in these switch statements.

* Source/WebCore/dom/Range.cpp:
(WebCore::processContentsBetweenOffsets):
(WebCore::Range::checkNodeOffsetPair):
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::appendNonElementNode):
* Source/WebCore/editing/MarkupAccumulator.h:
(WebCore::MarkupAccumulator::endAppendingNode):

Adopt dynamicDowncast<> while here.

Canonical link: https://commits.webkit.org/274207@main
